cmd/fossa/test.go

```diff
@@ -0,0 +1,127 @@
+package main
+
+import (
+	"errors"
+	"net/url"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/briandowns/spinner"
+	logging "github.com/op/go-logging"
+	"github.com/urfave/cli"
+	emoji "gopkg.in/kyokomi/emoji.v1"
+
+	"github.com/tidwall/gjson"
+)
+
+var testLogger = logging.MustGetLogger("test")
+
+const PollRequestDelay = 8000
+
+func confirmBuild(config cliConfig, timing int) error {
+	fossaBaseURL, err := url.Parse(config.endpoint)
+	if err != nil {
+		return errors.New("invalid FOSSA endpoint")
+	}
+
+	reqRef, _ := url.Parse("/api/revisions/" + url.PathEscape(config.getVcsLocator()) + "/build")
+	reqURL := fossaBaseURL.ResolveReference(reqRef).String()
+
+	testLogger.Debugf("Querying <%#v>", reqURL)
+	resp, err := makeAPIRequest("PUT", reqURL, nil, config.apiKey)
+	if err != nil {
+		return err
+	}
+
+	buildData := string(resp)
+	buildStatus := gjson.Get(buildData, "task.status").String()
+
+	testLogger.Debugf("Build status returned: %s", buildStatus)
+
+	switch buildStatus {
+	case "":
+		return errors.New("unable to parse build results")
+	case "FAILED":
+		return errors.New("failed to analyze build #" + gjson.Get(buildData, "id").String() + " <" + gjson.Get(buildData, "error").String() + ">; visit FOSSA or contact support@fossa.io")
+	case "SUCCEEDED":
+		return nil
+	default:
+	}
+
+	if timing >= config.timeout {
+		return errors.New("request series timed out")
+	}
+
+	time.Sleep(time.Duration(PollRequestDelay) * time.Millisecond)
+
+	return confirmBuild(config, timing+PollRequestDelay)
+}
+
+func confirmScan(config cliConfig, timing int) error {
+	fossaBaseURL, err := url.Parse(config.endpoint)
+	if err != nil {
+		return errors.New("invalid FOSSA endpoint")
+	}
+
+	reqRef, _ := url.Parse("/api/revisions/" + url.PathEscape(config.getVcsLocator()))
+	reqURL := fossaBaseURL.ResolveReference(reqRef).String()
+
+	testLogger.Debugf("Querying <%#v>", reqURL)
+	resp, err := makeAPIRequest("GET", reqURL, nil, config.apiKey)
+	if err != nil {
+		return err
+	}
+
+	revisionData := string(resp)
+	if gjson.Get(revisionData, "meta.0.last_scan").String() == "" {
+		// not scanned yet
+		if timing >= config.timeout {
+			return errors.New("request series timed out")
+		}
+		time.Sleep(time.Duration(PollRequestDelay) * time.Millisecond)
+
+		return confirmScan(config, timing+PollRequestDelay)
+	}
+
+	issueCount := 0
+	gjson.Get(revisionData, "issues.#.resolved").ForEach(func(key, value gjson.Result) bool {
```

Comment:
This is not JavaScript. Unmarshal this into a struct, and write a for loop over the list of issues.

Reply:
Why? This is effectively what gjson will do, but I believe gjson will provide some level of safety in case entries don't exist.

Reply:
What? The documentation for `gjson` says that behaviour is undefined when the input is malformed. No panic or error. You have _no idea_ what `gjson` will do.

Reply:
`gjson` is the opposite of safe. Here is the exact literal comment on `gjson.Get`:

```
// This function expects that the json is well-formed, and does not validate.
// Invalid json will not panic, but it may return back unexpected results.
```